aptos-move/block-executor/src/scheduler_v2.rs

```diff
@@ -66,3 +422,1400 @@ impl ExecutionQueueManager {
+        self.txn_statuses.start_executing(txn_idx)
+    }
+
+    /// Attempts to advance the `executed_once_max_idx` watermark.
+    ///
+    /// This watermark tracks the highest contiguous transaction index `i` such that all
+    /// transactions `0..i` have been executed at least once.
+    ///
+    /// This method is typically called when a transaction `txn_idx` finishes its 0-th incarnation.
+    /// If `executed_once_max_idx` is currently equal to `txn_idx`, it means `txn_idx` might
+    /// be the transaction completing a contiguous block of first-time executions.
+    /// The method then iterates from `txn_idx` upwards, checking `[ExecutionStatuses::ever_executed]`
+    /// for each subsequent transaction `idx`. For each `idx` that has been executed at least once:
+    /// - `executed_once_max_idx` is updated to `idx + 1`.
+    /// - If `idx` is found to be `[ExecutionStatuses::pending_scheduling_and_not_stalled]`, it is re-added to the
+    ///   `execution_queue`. This handles cases where a transaction `idx` might have been
+    ///   scheduled for re-execution (e.g., its first re-execution) but was deferred because
+    ///   `executed_once_max_idx` was less than `idx`. Now that the watermark is advancing
+    ///   past `idx`, it can be truly scheduled.
+    /// The iteration stops when an `idx` is encountered that has not `ever_executed` or when
+    /// `num_txns` is reached.
+    fn try_increase_executed_once_max_idx(&self, txn_idx: TxnIndex) {
+        let execution_queue_manager = &self.txn_statuses.get_execution_queue_manager();
+        // Similar to try_commit, synchronization is provided by the ordering of finish
+        // execution updating the transaction inner status (under lock), and the ever_executed
+        // check below, which also acquires the lock. In particular, ordering is as follows:
+        // (a) finish_execution(idx) with idx lock -> executed_idx == txn_idx check
+        // (b) increment executed_idx to txn_idx -> ever_executed check under lock
+        // Note that (classic flags principle), in case when ever_executed check fails,
+        // executed_idx == txn_idx check is guaranteed to succeed.
+        if execution_queue_manager
+            .executed_once_max_idx
+            .load(Ordering::Relaxed)
+            == txn_idx
+        {
+            let mut idx = txn_idx;
+            while idx < self.num_txns && self.txn_statuses.ever_executed(idx) {
+                // A successful check of ever_executed holds idx-th status lock and follows an
+                // increment of executed_idx to idx in the prior loop iteration. Adding stall
+                // can only remove idx from the execution queue while holding the idx-th status
+                // lock, which would have to be after ever_executed, and the corresponding
+                // remove_stall would hence acquire the same lock even later, and hence be
+                // guaranteed to observe executed_idx >= idx. TODO: confirm carefully.
```

Comment:
I found the comments in this method confusing.

Reply:
makes sense, I'll do a pass.

Reply:
did a pass, can you please re-check?